presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java

```diff
@@ -1062,6 +1064,52 @@ public void testOutputNotRequired()
         }
     }
 
+    @Test
+    public void testAdaptiveBatchSizes()
+            throws Exception
+    {
+        Type type = VARCHAR;
+        List<Type> types = ImmutableList.of(type);
+
+        TempFile tempFile = new TempFile();
+        List<String> values = new ArrayList<>();
+        int rowCount = 10000;
+        int longStringLength = 5000;
+        Random random = new Random();
+        for (int i = 0; i < rowCount; ++i) {
+            if (i < MAX_BATCH_SIZE) {
+                StringBuilder builder = new StringBuilder();
+                for (int j = 0; j < longStringLength; ++j) {
+                    builder.append(random.nextInt(10));
+                }
+                values.add(builder.toString());
+            }
+            else {
+                values.add("");
+            }
+        }
+        writeOrcColumnsPresto(tempFile.getFile(), DWRF, NONE, Optional.empty(), types, ImmutableList.of(values), new OrcWriterStats());
+
+        try (OrcSelectiveRecordReader recordReader = createCustomOrcSelectiveRecordReader(tempFile, OrcEncoding.DWRF, OrcPredicate.TRUE, type, MAX_BATCH_SIZE, false)) {
+            assertEquals(recordReader.getFileRowCount(), rowCount);
+            assertEquals(recordReader.getReaderRowCount(), rowCount);
+            assertEquals(recordReader.getFilePosition(), 0);
+            assertEquals(recordReader.getReaderPosition(), 0);
+
+            // Size of the first batch should equal to the initial batch size (set to MAX_BATCH_SIZE)
+            Page page = recordReader.getNextPage();
+            assertNotNull(page);
+            page = page.getLoadedPage();
+            assertEquals(page.getPositionCount(), MAX_BATCH_SIZE);
+
+            // Size of the second batch should be adjusted based on maxCombinedBytesPerRow collected during the first batch read
```

Comment:
perhaps, read one more time and and assert the batch size for the 3rd batch as well.

Reply:
Thanks, addressed comments.